### DISCUSSION TITLE
[API Request] Officialise POST support

### DISCUSSION THREAD

**Discussion by Tolriq** [opensubsonic/open-subsonic-api#26] [API requests]
### Type of change

API Clarification

### Proposal description

It seems quite a few servers already support that but we need to officialise POST support to the API to allow large batch queries like editing playlists, or future extensions.

### Backward compatibility

- [X] No backward compatibility impact.

### API details

To be documented by servers that already support that as we should use what's existing to ensure compatibility.

**Comment by Tolriq** [comment]
@opensubsonic/servers We need to move on with this to expand on the new APIs.

The server who already support some cases like form post or something else please post here and let's try to define something simple.

If there's no consensus for the new API it would be best to support json post and not form based list of params.

**Comment by deluan** [comment]
POSTs are accepted by Navidrome, based on a request by (IIRC) @MichaelBechHansen for play:Sub. It accepts a `x-www-form-urlcoded` payload, like this:

```bash
curl -v -X POST -H 'Content-Type: application/x-www-form-urlencoded' 'http://localhost:4533/rest/ping.view' --data 'c=client&v=1.8.0&f=json&u=user&p=password'
```

**Comment by gravelld** [comment]
POSTs are accepted by @Astiga for all requests. The query string can be used for parameters, or `x-www-form-urlcoded` can be used.

**Comment by sentriz** [comment]
gonic works the same as Navidrome and Astiga :+1: 

![image](https://github.com/opensubsonic/open-subsonic-api/assets/6832539/018b9564-d789-4801-b465-7dcec33ebd49)

**Comment by Tolriq** [comment]
Ok so seems @epoupon LMS does not yet support that, so we can't just say it's in OS and need an extension.

Since most already support `x-www-form-urlcoded` the extension will just say that this is supported so clients knows they can use it.

We can still add json post later as it won't be incompatible if server check the content type.

**Reply by Tolriq** [reply]
https://github.com/opensubsonic/open-subsonic-api/pull/74

**Comment by crazygolem** [comment]
The specification is too ambiguous regarding the handling of duplicated parameters in both the query and body, and specifically the authentication parameters.

The ambiguity could potentially lead to a parameter pollution vulnerability, where a component authenticates one user, and another component handles the request for another user, especially since query parameters and form parameters are often merged in library/framework-specific ways.

Although this could in principle be an issue already when you just have a subsonic server, it can actually become problematic in deployments where a reverse proxy in front is in charge of the authentication. It is especially problematic since a fully secure authentication service becomes potentially vulnerable when the subsonic backend is switched for an opensubsonic one that supports this extension.

I suggest to improve the specification to address the issue specifically for the authentication parameters, as ambiguity with the other parameters should not create a security risk.

My preferred solution would be to specify that authentication parameters must not appear in the body. If an alternative to query parameters is desired, the `Authorization` header is much more appropriate. This does not actually solve the issue entirely (you still have ambiguity between sources), but we avoid the issue of libraries/frameworks automatically merging parameters from multiple sources in different ways (which is the biggest footgun in this story).

Some other ideas but they all feel just off the mark:

* Servers should check each source independently and raise an error if parameters are found in several sources.
    * This is likely to break current implementations (e.g. clients sending basicauth credentials in addition to query parameters instead of replacing them, servers supporting the `Remote-User` header but not requiring the authentication service to remove the query parameters).
    * An alternative with lower risk of breaking things would be to specify an order between the sources, and in particular I suggest that any source specified by OS should have lower priority than query, to avoid the scenario where an authenticating reverse-proxy setup becomes vulnerable when the subsonic backend adds support for extra OS-specified sources. Custom sources can still have higher priority than query.
* Authentication parameters should all be fetched from the same source (either all from the query, all from the body, or all from a custom source), i.e. there should not be a fallback mechanism on individual authentication parameters.
    * This does not actually prevent the issue but makes it easier to implement checks.

What do you think?

**Reply by Tolriq** [reply]
Just to be clear most servers already works like this, the specification just expose a field so that client knows those who do or not for sure. (But all up to date did that before).

There's no change in anything for OS and your problems are not really tied OS but to some kind of potential proxy that would try to hijack subsonic API to make it authenticate to something else.

External authentification should be done via the Subsonic server calling that auth or via a mechanism specific to that solution like an auth header or other specific ways.

Just remember that if OS switch to Oauth2 or other more complex auth system your solution will probably not work anymore.

**Reply by crazygolem** [reply]
> Just to be clear most servers already works like this, the specification just expose a field so that client knows those who do or not for sure. (But all up to date did that before).

Yes, and it's not fun to learn after you have deployed your system that your subsonic server supports an undocumented feature that is not part of the specification and that makes your system potentially vulnerable. True story, for me it was a few days ago learning that navidrome supports this extension.

If a server is vulnerable because it does things on its own, that's one thing, I can switch servers if it comes to it. If a specification mandates servers to implement something potentially vulnerable, it's another thing entirely. In my opinion it's better to remove it again from the specification (servers can still do their thing as before; and if they don't document it or do it wrong it's on them) than to have a broken specification.

> There's no change in anything for OS and your problems are not really tied OS but to some kind of potential proxy that would try to hijack subsonic API to make it authenticate to something else.

From a security perspective, a specification is meant to be used for every component involved, not just the client and server. This includes use-cases such as introspection, DoS countermeasures, and authentication too (which is my use-case). As such, relying on a solid subsonic or opensubsonic specification to implement authentication in a reverse proxy is not "hijacking" it, but a valid use of the spec.

> Just remember that if OS switch to Oauth2 or other more complex auth system your solution will probably not work anymore.

If you don't see the problem with the way `postForm` is specified, you will likely specify the new auth mechanism in a similarly broken way, because the vulnerability doesn't depend on specificity of the mechanisms, but on the plurality of mechanisms...

... with a special caveat for the combination of query parameters and form parameters that together kind of get vulnerable by default for historic reasons. My suggestion to not allow authentication parameters in the POST body removes this caveat, and brings the number of official mechanisms back to one so the ambiguity is resolved.

Before the `formPost` extension was specified, servers were already not implementing the spec. If the spec is updated to prevent the vulnerability, server can be updated to implement the more secure version of the spec.

> (bis) Just remember that if OS switch to Oauth2 or other more complex auth system your solution will probably not work anymore.

The "switch" in "if OS  switch to OAuth2" is doing the heavy lifting here, and switching would be an entirely different situation than just adding the extra authentication mechanism. If OS "switched" to `formPost` instead of adding it as an alternative, the whole issue would go away (as far as the OS spec is concerned; for servers adding their own custom mechanism, it's on them). And if OS actually switched to a modern authentication (which I hope it does, but somehow doubt it will, to preserve backwards compatibility) it would indeed break my system **but it would not make it vulnerable**. And I could rely on a solid spec to update my system.

**Reply by Tolriq** [reply]
The auth in Subsonic is not secure by itself, this extension is not related to auth and there's plans to build a proper auth system.

Preventing multiple user and password is not enforced in the Subsonic API and you can already pass multiple times the get parameters. What servers does with that is undefined for now as it is with this extension.

This is not something to be defined in this extension as it touch also the get part ....

So still disagree with you, there's need for a proper auth system, but your request does not change anything about current state of Subsonic security with or without the extension.

**Reply by crazygolem** [reply]
> The auth in Subsonic is not secure by itself

The auth in subsonic is cryptographically weak and has many problems, but calling subsonic "not secure" is a bit of a stretch or at least too vague for a serious discussion. As an example, if it really was "not secure" you could simply get rid of authentication altogether in the opensubsonic spec even before a modern auth gets specified, and it would not make subsonic "less secure" (it would make it "more secure" because it would finally let servers hash passwords, yay).

Please don't use the weaknesses of the subsonic auth spec as an excuse to make the situation worse. Instead take the weaknesses into account when specifying extensions.

> this extension is not related to auth

Security doesn't happen in isolation, this extension is "related to auth" because it touches something that auth touches: how credentials are passed around.

> and there's plans to build a proper auth system

I'm aware and following with attention. The issue I'm raising now will not go away just with an extra authentication mechanism.

> Preventing multiple user and password is not enforced in the Subsonic API and you can already pass multiple times the get parameters. What servers does with that is undefined for now as it is with this extension.

Yes we agree on that, the vanilla subsonic spec is also not ideal, and I would much prefer that opensubsonic addresses the security security issues with vanilla subsonic rather than add its own.

> This is not something to be defined in this extension as it touch also the get part ....

My point is that adding this extension is not just adding another layer with the same issue as in the original spec that can be handled in the same way, rather the issue becomes much worse because of the way libs and frameworks historically handle the ambiguity between query parameters and form parameters. It's not just the same vulnerability, it's a brand new footgun.

**Reply by crazygolem** [reply]
But maybe I'm not explaining the issue in the right way, or coming at the argument from the wrong side.

For me the solution I'm proposing is straightforward to specify, straightforward to implement, and since we're already talking about extending the original subsonic spec to support form parameters, my assumption is that adding an extra sentence to the spec to exclude auth parameters should not be a big deal.

Servers are still free to read credentials from the form parameters if they want: it will become a bug according to the spec, but it won't break anything from what they currently do, they can take their time to fix the bug or chose not to; there is such a thing as "not fully compliant" if it doesn't break anything for their users.

Maybe you can explain why you think it's important to allow auth parameters in the POST form too?

**Reply by Tolriq** [reply]
What's important is to not break current servers to enforce things they maybe can not enforce.
OS goal is that when it says X servers do X to avoid current status of Subsonic servers.

So if we change postForm to say refuse auth there all current servers that currently expose this would need to remove the extension as would not be compliant. There's no half compliant in OS, and as you said many libraries can handle this for them and they would not be able to manage this properly without a ton of code.

Furthermore as I said, multiple auth param via GET is undefined, so allowing them via post does not change anything to current status of multiple auth values, passing then via GET or POST does not change anything from security POV here. 

You still have no shown any security impact of the current situation except saying that it might eventually cause an issue in the tool that you develop ....
You can enforce whatever security you want in the proxy you code to prevent what you do not want.

And for Subsonic auth, half the servers do not support the new auth system so it's not about crypto levels it's about having the login and password in CLEAR TEXT and for many users over http since they don't know how to properly secure things.

So yes Subsonic auth is not secure, specially when dealing with users that have no idea what secure is.

**Reply by crazygolem** [reply]
Is it important to you that the opensubsonic spec remains compatible with the subsonic spec?

By "remain compatible", I mean that any software that supports and expects the subsonic protocol should be able to handle the opensubsonic protocol too? E.g. opensubsonic client talks to a subsonic server, subsonic client talks to an opensubsonic server, and maybe some middleware that expects the subsonic protocol should not break with either subsonic or opensubsonic client or server?

> You still have no shown any security impact of the current situation except saying that it might eventually cause an issue in the tool that you develop.

What exactly do you expect me to show here?

I have explained in details the kind of vulnerability this spec introduces **at the spec level**, and why this particular combination of sources is worse than just one or the other. Is this not enough?

It actually caused an issue in the middleware I develop, because I developed it to support the subsonic protocol, not some badly specified extension, and not being aware that my subsonic backend already supported this extension. As a result, my server was actually vulnerable.

My middleware is safe against this vulnerability assuming the subsonic protocol is used; my assumption for opensubsonic would be that an officially specified extension **does not break things for anyone**.

I will obviously fix my middleware, it doesn't mean the spec is good or shouldn't be also fixed. Ideally, people should not be forced to use my tool if they want to do reverse-proxy auth, they should be able to rely on a solid specification that nudges servers and reverse-proxies into implementing sensitive stuff correctly.

> And for Subsonic auth, half the servers do not support the new auth system so it's not about crypto levels it's about having the login and password in CLEAR TEXT

So you're saying that there is a server that doesn't support the new auth system (I'm assuming you're talking about `t` + `s`?) but does support `postForm`?

From a security perspective, there is not much difference between having the password in clear text in the query or in the body...

> many libraries can handle this for them and they would not be able to manage this properly without a ton of code.

I think this is the heart of your argument: convenience over security.

I think that most devs that already implemented a subsonic server would not shy away from a few extra lines of code (it shouldn't be *a ton* as you say) *if they have a good spec to tell them what they are supposed to do*. Also, as far as I know, you are in the confortable position that servers that support opensubsonic, and specifically this fairly recent extension, are actively maintained, so can adapt.

Or if you still think that the number of lines of code really matter, consider that my security (and authentication) middleware will need far many more "tons" of lines of code to properly handle this spec in all its ambiguous glory: not only do I have to do extra checks to make sure the auth parameters can never be interpreted ambiguously under any circumstances, but I also need to rewrite the body (for the authentication part: the credentials should not be forwarded to the backend if the proxy handles that) which can open up my server to a bunch of new vulnerabilities (body rewriting is more risky than url rewriting). And then (speculation) you add another similar extension for using JSON in the body, and I'm screwed all over again, I have to introduce another subsystem to handle JSON instead of form parameters. ALL THAT is an awful ton of extra code, and don't get me wrong, I don't mind writing tons of code when necessary, but I don't like writing tons of code that become vulnerable as soon as a new shitty extension is produced >:-(

**Reply by crazygolem** [reply]
Hey @opensubsonic/servers, if you support this extension, could I have your input on how difficult it would be for you to not accept auth parameters in the body?

**Reply by sentriz** [reply]
@crazygolem i haven't been following the discussioin here closely. but i imagine it's quite easy for servers to ignore auth params in just forms 

<details>
<summary>eg  (gonic)</summary>

```diff
diff --git a/server/ctrlsubsonic/params/params.go b/server/ctrlsubsonic/params/params.go
index 022c4e2..adbf14e 100644
--- a/server/ctrlsubsonic/params/params.go
+++ b/server/ctrlsubsonic/params/params.go
@@ -45,6 +45,10 @@ func New(r *http.Request) Params {
 	// also if there's any in the post body, use those too
 	if err := r.ParseForm(); err == nil {
 		for k, v := range r.Form {
+			switch k {
+			case "u", "t", "s", "p":
+				continue
+			}
 			params[k] = v
 		}
 	}
```

</details>

**Reply by Tolriq** [reply]
You mix things, what is important is to not change the current API definition that would break existing clients.
Servers already support POST, clients already use POST. If you change how POST works then you break all legacy clients that rely on POST and pass the auth params there.

All that to fix your very specific proxy with 0 security gain for the vast majority of people?

So again this extension does not change anything, it just allow clients to know what server support post without ANY BREAKING CHANGE.
For the record it's actually all the servers ....

Now about the code, I don't care about code, I care about server work, seems for Gonic it's simple, now imagine I use a library that just callback my code with all the params without distinction from where they come? I need to rewrite the server to use another library? I need to write myself a webserver? I need to manually parse everything? You have absolutely no idea about the impact on server code this simple change can make. (And again ignoring that it can break a few legacy clients ...)

And yes there's OS servers that do not support the new auth because the new auth is dumb and requires storing the password in clear and does not allow forwarding of the auth to external more secure systems....

And yes clear text in body or not have 0 impact as I said too, and GET already does not specify what to do if you have multiple u and p parameters.

From Subsonic POV that does not change anything about security to allow or not the auth params in the headers or in the post or in both.

TL;DR; your request brings nothing to Subsonic servers including security, but breaks legacy clients as all servers already support POST it would force all servers to update their code for your proxy.

PS: A lot of clients support adding headers for proper authentification without the need for special proxy that tries to do strange things over Subsonic API. API is clearly not well defined and the reason a new one is created.

So give your requirements in the new auth thread, but we're not breaking legacy clients over your very specific proxy.

**Reply by sentriz** [reply]
just to add, and could be wrong, the only client that I know of that uses form params is play:Sub. so we could check if it uses that for auth or not. if not then there's no issue presumably @MichaelBechHansen

**Reply by Tolriq** [reply]
Since POST is the only way to properly manage large playlists I'm pretty sure others have discovered and used that since then.

**Reply by crazygolem** [reply]
> Servers already support POST, clients already use POST. If you change how POST works then you break all legacy clients that rely on POST and pass the auth params there.

Supporting POST and supporting *the `formPost` extension* are not exactly the same thing: legacy clients don't support `formPost`, they just throw whatever at the server hoping it works (it's an image), legacy servers don't support `formPost`, they just go "meh whatevs" (also not literal).

This being an *extension*, I was assuming a new thing by opensubsonic for servers that actually support the opensubsonic spec, and the way the documentation has been written in the [API Reference](https://opensubsonic.netlify.app/docs/api-reference/#post-support) suggests to me that it was actually intended to be that way, where actually only the part about servers letting clients know is the extension.

If legacy clients and servers are involved, the basic support for POST form parameters should probably be rebranded as a *clarification* instead, according to the terminology on [OpenSubsonic changes](https://opensubsonic.netlify.app/docs/opensubsonic-changes/).

With that, I'll drop the idea of not allowing authentication parameters in the body, and instead we can do the following:

1. Clarify that all parameters should be located either in the query **or** the body, not both.
    * Legacy and actively maintained clients likely already do that, there is not a good reason to mix both sources if you're not trying to break the server.
    * Legacy servers don't care about opensubsonic, so they won't be impacted: if they support post, they already support compliant clients. They will still be vulnerable, but that won't change from the situation before the clarification, also they are called "legacy" for a reason.
    * Actively maintained servers that support this extension can either assume clients will be nice, or add an extra check if they want to be on the secure side.
    * Middlewares can rely on a clear, security-conscious spec to implement their stuff securely.
2. Clarify that each authentication parameters should appear at most once
    * Legacy and actively maintained clients likely already do that, it's just the normal thing to do.
    * Legacy servers don't care: same as above, they don't care when the clients are compliant, they are potentially vulnerable when not but it doesn't change the status quo.
    * Actively maintained servers: same as above.
    * Middleware: same as above.
3. Maybe add a security warning to the documentation, that's just the nice thing to do, and is likely to be propagated to the servers' documentation when they support reverse proxies.

Note that:
* **The spec restricts clients, not servers**: if the client sends something unexpected, servers can deal with it however they want, as always. **Servers don't have to do anything if they don't want to and still remain compliant.**
* Genuine clients, and most importantly legacy clients, already work this way, simply because it's the normal thing to do. As such, **it doesn't break compatibility**.
* It doesn't actually solve the parameter pollution vulnerability entirely, but makes it easier to prevent against it because the remaining issue is highlighted in the spec for implementors.

I still think that the extension misses the mark, as legacy clients that do POST will still send POST to an opensubsonic server that does not support POST, and actively maintained clients that do support opensubsonic could benefit from a better (security-wise) extension because they are able to adapt.

As a clarification of the original subsonic spec, I don't think we need to change much to the extension's specification. But if needed, since the `formPost` extension is relatively new, I think there is a chance to have all clients and servers that support it agree to the change, and we can just add the clarifications to the extension. Otherwise, we could create a new extension with the updated spec, let's call it `formPost2`, and deprecate `formPost`.

What do you think about this?

**Reply by Tolriq** [reply]
Any change even as suggestion is breaking ....

And yes it's common to pass auth via get and still pass the rest via form specially when code is conditional to the server it talks to .... Many client have if based on server name.

Global auth interceptor adding auth and actual query code that is either get or post depending on the server ....

And for the last time no we will not change current auth spec for no reason as it's breaking. We will build a new one.

The spec is all about servers.... That's their job to ensure it's respected and enforce what it says ....
If you want clients to do something ask directly the client you use, if client have choice they will use what's best for them.
